[WIP] Fix behavior of "dim=()" and "dim=None" in reduction methods to match numpy

kurtamohler opened the pull request:
dim=() means that no reduction should occur, as in numpy

Issue #29137

kurtamohler committed:
Fix empty tuple dim argument handling in sum/prod/mean/norm

An empty tuple means that no reduction should occur

Issue #29137

kurtamohler committed:
Update var_mean and std_mean to handle dim=() properly

kurtamohler committed:
Add tests for reduction methods where dim=()

kurtamohler committed:
Clean up unnecessary snippets of code

ezyang commented:
Nice, you sidestepped the `int[2]?` problem very nicely in this PR :)

ezyang commented:
I'm a little uncertain if we can merge this directly though: not only is it a BC-breaking change, and with this PR only, there's no way to recover the old behavior. Although you could argue there is no reason to keep BC with a bug, we should *at very least* offer our users a way to explicitly ask for the old behavior, and there is no way to do so.

ezyang commented:
Substantive content of the PR LGTM; approval withheld due to BC concerns.

kurtamohler commented:
@ezyang, yeah that makes sense. If someone has existing code that uses `tensor.sum(dim=())`, this fix would mess them up.

kurtamohler commented:
@ezyang, so how should we allow the user to ask for the old behavior?

rgommers commented:
> there's no way to recover the old behavior

The old behavior is identical to `dim=None` right? So that's how to ask for it. It wouldn't really make sense to introduce some other mechanism to switch the behavior of `dim=()`, and then later remove that mechanism again (that would require users to change their code twice).

It seems to me like the choice is between first giving a `FutureWarning` for this change and waiting with the merge till a future release, or just putting it in as is now and adding a release note about it.

ezyang commented:
> The old behavior is identical to dim=None right? 

Yes, which this PR doesn't introduce. So it seems that we will have to shave the `dim[1]?` yak in any case.

rgommers commented:
> > The old behavior is identical to dim=None right?
> 
> Yes, which this PR doesn't introduce. So it seems that we will have to shave the `dim[1]?` yak in any case.

Ah, I somehow managed to miss that. Yes, that seems critical.

Commit:
Add partial support for int[]? optional argument

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 90e571ed:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=90e571ed01a5867e8591879abefe67c28a381529).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4121030) pytorch_linux_xenial_py3_6_gcc5_4_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79921734) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79921734))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py 
Auto-merging .circleci/generate_config_yml.py 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_travis_python.sh 
Auto-merging .circleci/docker/common/install_travis_python.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/common/install_conda.sh 
Auto-merging .circleci/docker/common/install_conda.sh 
CONFLICT (add/add): Merge conflict in .circleci/docker/build.sh 
Auto-merging .circleci/docker/build.sh 
CONFLICT (add/add): Merge conflict in .circleci/config.yml 
Auto-merging .circleci/config.yml 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4121031) pytorch_xla_linux_xenial_py3_6_clang7_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79923902) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79923902))
<details>
<summary>
<code>Jan 03 20:49:02 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff</code>
</summary>

```
Jan 03 20:49:02 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 03 20:49:02 1 out of 1 hunk FAILED 
Jan 03 20:49:02 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 03 20:49:02 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 03 20:49:02 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Jan 03 20:49:02 Traceback (most recent call last): 
Jan 03 20:49:02   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 03 20:49:02     patch_repo(args) 
Jan 03 20:49:02   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 03 20:49:02     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 03 20:49:02 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-serialization.diff 
Jan 03 20:49:02 + cleanup 
Jan 03 20:49:02 + retcode=1 
Jan 03 20:49:02 + set +x 
Jan 03 20:49:02 =================== sccache compilation log =================== 
Jan 03 20:49:02 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 03 20:49:02 Compile requests                 1 
Jan 03 20:49:02 Compile requests executed        0 
Jan 03 20:49:02 Cache hits                       0 
Jan 03 20:49:02 Cache misses                     0 
Jan 03 20:49:02 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

kurtamohler commented:
With this most recent change, I was able to set the dim argument to type `int[]?` instead of `int[1]` in native_functions.yaml and successfully build the project. However, my work is not complete for two reasons:

- If I set the dim argument to `int[1]?`, I get a build error
- When I try to run `sum(dim=None)`, I get this error:
```
Traceback (most recent call last):
  File "../pytorch-scripts/sums.py", line 18, in <module>
    a.sum(dim=None)
RuntimeError: Please look up dimensions by name, got: name = None.
```
It seems that when I set dim=None, pytorch is attempting to run the `sum.dim_DimnameList` version of the sum method, whereas I want it to run the `sum.dim_IntList` version. I'm not sure yet how to change this. If anyone knows, please let me know.

zou3519 commented:
> It seems that when I set dim=None, pytorch is attempting to run the `sum.dim_DimnameList` version of the sum method, whereas I want it to run the `sum.dim_IntList` version. I'm not sure yet how to change this. If anyone knows, please let me know.

Our PythonArgParser matches against schemas in some ordering order. In particular, the`sum.dim_DimnameList` schema comes before the `sum.dim_IntList` schema. I've known about this quirk for a couple of months but I don't know what the best way to fix it is.  I think this is orthogonal to the real concern though, which is getting the build to understand "int[1]?"

kurtamohler commented:
> I think this is orthogonal to the real concern though, which is getting the build to understand "int[1]?"

Ok, so you think it might call `sum.dim_IntList` properly if I can figure out how to get the `1` to be understood correctly?

zou3519 commented:
> > I think this is orthogonal to the real concern though, which is getting the build to understand "int[1]?"
> 
> Ok, so you think it might call `sum.dim_IntList` properly if I can figure out how to get the `1` to be understood correctly?

Hmm... So there are two problems here. The first is that we should try to get `int[1]?` to be understood. The second one is that the `sum.dim_DimnameList` schema appears before `sum.dim_IntList`.

We'll need to figure out both -- if we get the build to understand `int[1]?`, I'm not sure which overload will appear first. The rules for what order the schemas get generated in is hard to find, but I agree that it would be good to figure that out. I'll dig into the code generation a little next week, by next Wednesday (I might not have time to figure this out before then)

kurtamohler commented:
Ok then I'll focus on getting `int[1]?` working first. Also, I misspoke earlier. `int[1]?` doesn't give a build error. It builds, but when I try to run `import torch`, I get this error:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  expected ident but found '?' here:
aten::sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
                                         ~ <--- HERE
```
At the moment, I'm not sure how to figure out where exactly that error message happens or what it means

kurtamohler commented:
I ran a backtrace in gdb and found out that the error is happening in the `parseArgument()` function in `torch/csrc/jit/script/function_schema_parser.cpp`. I just need to figure out why it's failing.

kurtamohler committed:
Update sum dim type to 'int[1]?'

Allow trailing '?' after '[n]' in schema argument parser

kurtamohler commented:
My latest commit allows `int[n]?` to pass the parsing checks in `torch/csrc/jit/script/function_schema_parser.cpp`, so `import torch` works again. Now we just need to figure out the `sum.dim_IntList` versus `sum.dim_DimnameList` issue.

kurtamohler commented:
I just found another issue when I tried running test/test_torch.py. It seems to incorrectly handle cases where dim is set to an int:
```
>>> a.sum(dim=0)
TypeError: sum() received an invalid combination of arguments - got (dim=int, ), but expected one of:
 * (torch.dtype dtype)
      didn't match because some of the keywords were incorrect: dim
 * (tuple of names dim, bool keepdim, torch.dtype dtype)
 * (tuple of ints dim, bool keepdim, torch.dtype dtype)
```
This used to work before my recent changes. It doesn't seem like this should be too difficult to fix though.

kurtamohler commented:
I found the reason why setting dim to an int gives an error. In `csrc/autograd/generated/python_variable_methods.cpp`, the `THPVariable_sum()` function creates a PythonArgParser object, and it's setting the dim argument type in the schema to be `IntArrayRef?[1]`. This is incorrect--it should be `IntArrayRef[1]?`. When I fix it manually and rebuild, setting dim to an int works fine. However, this is a generated file, so I'll need to figure out how to fix the script that generates it, which is `tools/autograd/gen_python_functions.py`.

kurtamohler committed:
Fix how gen_python_functions.py parses arguments like "[x]?"

kurtamohler commented:
I've fixed the aforementioned issue.

But now I notice that there's an issue with sum_backward(). Here's what I see:
```
>>> a = torch.ones(10,5)
>>> a.requires_grad = True
>>> b = a.sum(dim=(0))
>>> b.backward(torch.ones_like(b))
Traceback (most recent call last):
  File "../pytorch-scripts/sum_backward.py", line 8, in <module>
    b.backward(torch.ones_like(b))
  File "/home/kurtamohler/development/pytorch-reduction/torch/tensor.py", line 195, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/kurtamohler/development/pytorch-reduction/torch/autograd/__init__.py", line 99, in backward
    allow_unreachable=True)  # allow_unreachable flag
IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 94207527462128)
```

The error is happening in the sum_backward() function defined in `tools/autograd/templates/Functions.cpp`, one of the functions that this PR touches. It looks like the value of the IntArrayRef held in the c10::optional opt_dim object is garbage. I added a std::cout to print out its value, and it does match the value reported by the IndexError. I guess this must mean that the IntArrayRef is either getting deleted or overwritten for some reason between the time of the sum() call and the sum_backward() call. Maybe it has something to do with how the c10::optional class holds onto the IntArrayRef value.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -403,6 +405,13 @@ def unpack_variable(name, unpack_expr, typename):
+            if typename == 'c10::optional<IntArrayRef>':
+                result = """\
+                    auto __{name} = {expr};
+                    c10::optional<{typ}> {name} = __{name} ? c10::make_optional({typ}(__{name}.value())) : c10::nullopt;
```
This is wrong. You're taking out a reference to `value()` but `__{name}.value()` is a temporary and will die once you exit this scope. That's why you're getting garbage.

kurtamohler replied:
Ok I see. According to a comment just above this code:
```
# optional<ArrayRef<T>>. One needs to unwrap the optional and rewrap.
```
I'm not sure how to do that.

ezyang commented:
Adding @zdevito for JIT changes

zdevito commented on torch/csrc/jit/script/function_schema_parser.cpp:
```diff
@@ -137,6 +137,7 @@ struct SchemaParser {
         name = "";
       }
     } else {
+      L.nextIf('?');
```
Adding support for optional sized arrays, e.g. `int[3]?`, will require more changes in the JIT than what are in this patch. In particular tryMatchArgument will need to be modified to correctly handle None values.

zdevito reviewed (commented):
I am confused why passing an empty list requires use to change the schema to accept an optional argument. In every test, the dim is passed as an empty list.

kurtamohler commented:
> I am confused why passing an empty list requires use to change the schema to accept an optional argument. In every test, the dim is passed as an empty list.

ezyang wants support for `dim=None`, in which case we would need the optional argument.

zdevito commented:
I read more of the thread... If the user has to change their code anyway to get the old behavior, isn't it possible to just choose _any_ other way to do it that would be easier to implement. The amount of changes to support `int[3]?` is a lot, and it changes the type system of TorchScript. If the only reason is to introduce a way to opt into old behavior manually, I'd prefer just `_legacy_dims=False` in argument list or something.

rgommers commented:
Note that the original decision in triage review was "lining up with NumPy". In NumPy code `axis=None` is extremely common, while `axis=()` is not. So changing the behavior of `dim=()` but skipping implementing `dim=None` for an easier alternative is not very useful. At that point, it's better to say that this is too much work and do nothing.

zdevito commented:
Oh, I see, that is much better context. It would be helpful to put that discussion in the PR description.  I feel like I still need more context here. Specifically, why is the reduction dimension described as `int[1]` in the first place?  If I had to describe, in natural language, what the intent is for these different types, I would do something like:

* `int[]` - a list of integers
* `int[4]` - a list of integers, or a single integer. If it was a single integer, replicate that integer 4 times.

The second one is used to represent places where padding, or other stuff might be specified for each dimension, but really is frequently the same for all dimensions. The use of `int[1]` in reductions is surprising because it doesn't fit the pattern that all the other uses of sized integers fit. Instead it is a trick to allow both 1 and [3, 4, 5] to appear as arguments. But now we also want to allow None because it makes sense for specifying a list of dimensions. It would be more orthogonal, and potentially easier to introduce a DimList type, which is an abstract type that accepts either a list of integers, a single int, or None.

kurtamohler commented:
That seems like a good approach, @zdevito. I can start adding a DimList type unless anyone has a reason not to.

zdevito commented:
@ezyang @bhosmer - curious as to your thoughts on if we should add DimList as an abstract type (i.e. one that appears in arguments but not in generally as a value). It seems to better capture the intent in the places where it is used. The downsides would be having to thread it through several places to get the behavior correct, but I feel like that would be simpler work that trying to deal with optional sizes integer arrays.

bhosmer commented:
> @ezyang @bhosmer - curious as to your thoughts on if we should add DimList as an abstract type (i.e. one that appears in arguments but not in generally as a value). It seems to better capture the intent in the places where it is used. The downsides would be having to thread it through several places to get the behavior correct, but I feel like that would be simpler work that trying to deal with optional sizes integer arrays.

Agreed, definitely seems better to encapsulate this to me. If by "thread it through" you just mean DimList would need to be understood wherever lvalue types are in play, sure, but presumably those same places would need to behave as desired regardless, right? A nominal DimList type would give the behavior an explicit locator to hang the behavior off of, that seems like a net simplification.

Couple questions

1. Are there use cases that would need DimList to be parameterized by an extent, `DimList[4]` etc.?

2. Would DimList muddy the waters at all for overload resolution?

ezyang commented:
@zdevito I think your observation about the replicate behavior, and how it's not applicable to the case of `dim` here, is spot on. But, to take your reasoning one step further, it's a bit awkward that we robbed the name `int[2]` to mean "list of integers, but if you have a single integer replicate it 2 times"; I probably would have reserved this name to mean, "two integers, exactly". And I also privately have an idea about my head of what the semantics of a `[n]` postfix modifier are, and it gets a bit spoopy when we talk about `DimList[2]` (in the case of `int[2]`, the postfix is some sort of array-ification, whereas with `DimList[2]`, DimList is parametrized by an integer and you're passing it through).

The other thing I'm concerned about, is that today our type system is very structural, and `ReductionDim` (I'm going to propose we call it this, rather than `DimList`) would be our first instance of a more nominal type in our type system. The trouble with doing nominal types is that we may end up having multiple nominally distinct types which share the same structure, and if we sort of just bash this through the existing code, I fear we'll end up duplicating code that deals with the structural aspects of the code. The risk is not too big with `ReductionDim` but it will get more acute if we add more nominal types of this sort. Also, if you design a sufficiently big structural type system, people can shoehorn their crap into your types without having to extend the type system (kind of like how you don't have to extend JSON, it's got all the generic crap you need).

ezyang commented:
I think this is probably a good enough idea that I'd be willing to at least try it, and see if it works well or not. (But with the knowledge that this will probably be the harbinger of many more types of this sort)

bhosmer commented:
@ezyang @zdevito Ed a few thoughts prompted by your points
1. Agree that installing nominal types as an idiom as a big step. I'd taken from Zach's original comments the idea that properly implementing the structural behaviors that we'd need to combine into the desired use case would be, like, massively harder. I guess it's worth quantifying the difference since, as you say, nominal types do risk turning into a get out of jail free card that gets abused.
2. Absolutely right about the semantic mismatch between `int[4]` and `NominalType[4]` - strictly speaking we'd want to introduce new parameterization syntax like `NominalType<4>` to avoid having the behavior switch on type names. Do we want to do that? Counterargument would be that `[]` already denotes something more tricky/context-dependent than `scalar-type[dims]`
3. `ReductionDims` does indeed feel like a step towards a big pile of very situation-specific nominal types... we could go in the opposite direction, and own the immediate purpose of avoiding generalized support of `int[n]?` by calling it like `OptionalIntList` or something, to flag the intent of replacing it with structural support later.

zdevito commented:
I wrote up my thoughts on the nominal types more completely below. I am still ambivalent. I feel like we would be better at the extremes -- either aggresively specify types for these special snowflakes as nominal types, or aggressively remove them and replace them with enough structural typing we do not need any of them (would require figuring out how to get rid of `int[3]`).

# Nominal Types to regularize Schema


## What? We have  number of places we have weird rules to try to address specific uses cases in schema:


* Per-dimension configuration int[3] Accept a list of integers, or a single integer. If a single integer, convert to a list of 4 integers. Used for non-square specification of padding, strides, dialation, etc. for convolution. Annoying because it means the type signatures of 1, 2 and 3D convolution are different
* int[1] Used in our reduction ops to specify a list of reduction dimensions. Also accepts a single int. However, we want to also allow the specification of None, meaning accept all the dimensions. 


The sized lists thing is annoying because it is not really a type, it is an annotation on a function schema specifying weird broadcasting behavior. It is hard to annotation in Python, has special handling in PythonArgParser and the JIT.

## Proposal
We replace these uses with types that explicitly represent the behavior we find in these cases. In particular, we have:

*  ReductionDim - replaces int[1] in the reduction cases, None, Int, and list of ints implicitly convert to ReductionDim
* Padding (I hate this name, someone can bikeshed it to something better)- replaces int[3], int[2]. etc. It can be a single int or a list of integers. The specific size is not encoded in the schema. Each op already knows the size that something can be and so it can cause the int to be replicated to the right size.

## How? 

* In C++: We introduce real classes for these, similar to Device, and add them to IValue.
* In JIT: We introduce types for these, and conversion rules as per above.
* In Python bindings: we change the glue code to implement the same rules.

In this process, we remove sized integers from schema as they were only ever used for the above two cases.

## Risks

* If we end up with more of these things, we might get a proliferation of types. However, these are long-standing things in the type system, similar to Device or Layout specifications.
* This is more work than I think it is.

zdevito commented:
This discussion is also related to TensorOptions, which is a major nominal type that we have special handling for. Decisions about TensorOptions and decisions on these should probably be made together.

ezyang commented:
cc @izdeby

ezyang commented:
I support @zdevito's plan. Instead of Padding let's call it PerDim (pun not intended).

bhosmer commented:
Yeah, +1 @zdevito @ezyang

kurtamohler commented:
@zdevito , I'm trying to add the ReductionDim class at the moment, and I have a few questions.

I chose to add it to the c10 namespace, but would it be better to add it to aten?
Are there any readme's or other documents that describe all the different file changes I'd need to make to get it working? I'm having a fair amount of trouble with my current strategy, which has been: attempt to build, make more changes to try to fix errors, repeat.

kurtamohler commented:
After some discussion with @ezyang, @rgommers has asked me to step down from this issue to let a maintainer work on it instead, since there are too many unknowns about adding the ReductionDim type.

I've pushed my latest changes for attempting to add ReductionDim to a separate branch. Obviously, I wasn't able to get it working, but I made it available in case some of it is useful. https://github.com/kurtamohler/pytorch/tree/reduction-dim-empty-tuple-29137-ReductionDim-type-unfinished

kurtamohler commented:
Replaced by PR #43982
